File: android/app/src/main/java/gg/yappy/app/ui/components/PixelPet.kt
package gg.yappy.app.ui.components

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.runtime.State
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import kotlin.math.abs

/**
 * The group pet: a pixel creature, drawn from character grids, animated by
 * flipping between two frames the way a Tamagotchi did.
 *
 * Sprites are code, not assets — each frame is sixteen strings of sixteen
 * characters, one character per pixel, so the art is diffable, palette-swaps
 * with the group's identity colour, and never ships a PNG. The vocabulary:
 *
 *   .  transparent      o  outline          b  body (identity colour)
 *   B  body shade       w  white            p  pink (tongue, inner ear)
 *   e  eye              y  brand yellow (sparkles, crown)
 *
 * Species is derived from the conversation id: half the world's groups get a
 * dog, half a cat, nobody chooses and everybody agrees.
 */

enum class PetSpecies { Dog, Cat }

fun petSpecies(conversationId: String): PetSpecies =
    if (conversationId.hashCode() and 1 == 0) PetSpecies.Dog else PetSpecies.Cat

// ─── Sprites ─────────────────────────────────────────────────────────────────

private val EGG = listOf(
    listOf(
        "................",
        "................",
        "......oooo......",
        ".....obbbbo.....",
        "....obbwbbbo....",
        "...obbwbbbbbo...",
        "...obbbbbbbbo...",
        "..obbbbbbbbbbo..",
        "..obbBbbbbBbbo..",
        "..obbbbbbbbbbo..",
        "..obBbbbbbbBbo..",
        "...obbbbbbbbo...",
        "....oobbbboo....",
        "......oooo......",
        "................",
        "................",
    ),
    listOf(
        "................",
        "................",
        "................",
        "......oooo......",
        ".....obbbbo.....",
        "....obbwbbbo....",
        "...obbwbbbbbo...",
        "...obbbbbbbbo...",
        "..obbbbbbbbbbo..",
        "..obbBbbbbBbbo..",
        "..obbbbbbbbbbo..",
        "..obBbbbbbbBbo..",
        "...obbbbbbbbo...",
        ".....oooooo.....",
        "................",
        "................",
    ),
)

// Dog: floppy ears, big muzzle. Frame two lifts the ears and wags.
private fun dogFrames(mood: String): List<List<String>> {
    // Rows 10-11 are the mouth region, swapped per mood.
    fun body(earUp: Boolean, mouthA: String, mouthB: String): List<String> {
        val e1 = if (earUp) "..oo........oo.." else "................"
        val e2 = if (earUp) ".obbo......obbo." else "..oo........oo.."
        val e3 = if (earUp) ".obBbo....obBbo." else ".obbo......obbo."
        return listOf(
            "................",
            e1,
            e2,
            e3,
            ".obBbooooooBbbo.",
            ".obbobbbbbbobbo.",
            "..oobbbbbbbboo..",
            "..obbebbbbebbo..",
            "..obbbbbbbbbbo..",
            "..obbbBooBbbbo..",
            mouthA,
            mouthB,
            "...obbbbbbbbo...",
            "....oooooooo....",
            "................",
            "................",
        )
    }
    return when (mood) {
        "happy" -> listOf(
            body(false, "..obbboppobbbo..", "...obbboppbbo..."),
            body(true, "..obbboppobbbo..", "...obbbbppbo...."),
        )
        "hungry" -> listOf(
            body(false, "..obbbboobbbbo..", "...obbbbbbbbo..."),
            body(false, "..obbboooobbbo..", "...obbbbbbbbo..."),
        )
        else -> listOf( // sad
            body(false, "..obbbboobbbbo..", "...obbboobbbo..."),
            body(false, "..obbbboobbbbo..", "...obbboobbbo..."),
        )
    }
}

// Cat: pointed ears, small mouth. Frame two flicks an ear and the tail.
private fun catFrames(mood: String): List<List<String>> {
    fun body(flick: Boolean, mouthA: String, mouthB: String): List<String> {
        val e1 = if (flick) "..o..........o.." else "..o.........o..."
        val e2 = if (flick) "..opo......opo.." else "..opo......opo.."
        return listOf(
            "................",
            e1,
            e2,
            "..obpo....obpo..",
            "..obboooooobbo..",
            ".obbbbbbbbbbbbo.",
            ".obebbbbbbbebbo.",
            ".obbbbbbbbbbbbo.",
            "..obbBwbbwBbbo..",
            mouthA,
            mouthB,
            "..obbbbbbbbbbo..",
            "...obbbbbbbbo...",
            "....oooooooo....",
            "................",
            "................",
        )
    }
    return when (mood) {
        "happy" -> listOf(
            body(false, "..obbbopbobbbo..", "...obbbbbbbbo..."),
            body(true, "..obbbobpobbbo..", "...obbbbbbbbo..."),
        )
        "hungry" -> listOf(
            body(false, "..obbbboobbbbo..", "...obbbbbbbbo..."),
            body(true, "..obbbboobbbbo..", "...obbbbbbbbo..."),
        )
        else -> listOf( // sad
            body(false, "..obbboBBobbbo..", "...obbbbbbbbo..."),
            body(false, "..obbboBBobbbo..", "...obbbbbbbbo..."),
        )
    }
}

// Wandered off: an empty spot — footprints trailing away and a question.
private val GONE = listOf(
    listOf(
        "................",
        "......ww........",
        ".....w..w.......",
        "........w.......",
        ".......w........",
        ".......w........",
        "................",
        ".......w........",
        "................",
        "..BB............",
        "..BB....BB......",
        "........BB......",
        "............BB..",
        "............BB..",
        "................",
        "................",
    ),
)

// Elder crown, drawn over the sprite's head rows.
private val CROWN = listOf(
    "....y..y..y.....",
    "....yyyyyy......",
)

// ─── Rendering ───────────────────────────────────────────────────────────────

/**
 * @param conversationId decides species and the body colour.
 * @param stage egg | baby | kid | grown | elder — size and form.
 * @param mood happy | hungry | sad | gone — expression and tempo.
 */
@Composable
fun PixelPet(
    conversationId: String,
    stage: String,
    mood: String,
    size: Dp,
    modifier: Modifier = Modifier,
    animated: Boolean = true,
) {
    val species = petSpecies(conversationId)
    val body = colorForId(conversationId)
    val shade = Color(
        red = body.red * 0.72f,
        green = body.green * 0.72f,
        blue = body.blue * 0.72f,
        alpha = 1f,
    )

    val frames = when {
        mood == "gone" -> GONE
        stage == "egg" -> EGG
        species == PetSpecies.Dog -> dogFrames(mood)
        else -> catFrames(mood)
    }

    // Sad pets breathe slowly; happy ones can barely sit still.
    val periodMs = when (mood) {
        "happy" -> 380
        "hungry" -> 650
        else -> 900
    }

    /*
     * The clock, kept away from composition.
     *
     * The first version read `phase` in the composition scope, which
     * invalidated every visible pet's composition at 60fps, forever — the
     * home list literally never went idle. Three changes hold it down:
     *   • a pet with nothing to animate gets no transition at all;
     *   • the two-frame flip goes through derivedStateOf, so composition
     *     only invalidates when the frame actually changes (~2/sec);
     *   • the bob offset is read inside the Canvas draw lambda, which is a
     *     repaint, not a recomposition.
     */
    val bobActive = animated && mood == "happy" && stage != "egg"
    val animate = animated && (frames.size > 1 || bobActive)
    val phase: State<Float> = if (animate) {
        rememberInfiniteTransition(label = "pet").animateFloat(
            initialValue = 0f,
            targetValue = 2f,
            animationSpec = infiniteRepeatable(
                animation = tween(periodMs * 2, easing = LinearEasing),
                repeatMode = RepeatMode.Restart,
            ),
            label = "petFrame",
        )
    } else {
        remember { mutableStateOf(0f) }
    }
    val frame by remember(frames, animate) {
        derivedStateOf { if (animate && frames.size > 1 && phase.value >= 1f) 1 else 0 }
    }

    val grid = frames[frame]
    // Babies are the same creature, smaller in the same box.
    val scale = when (stage) {
        "baby" -> 0.72f
        "kid" -> 0.88f
        else -> 1f
    }

    // Each grid resolved to coloured cells once — the draw pass used to walk
    // all 256 characters and re-answer the palette per cell, per frame.
    val bodyCells = remember(grid, body, shade) { resolveCells(grid, body, shade) }
    val crown = stage == "elder" && mood != "gone"
    val crownCells = if (crown) remember(body, shade) { resolveCells(CROWN, body, shade) } else null

    Canvas(modifier.size(size)) {
        val cells = 16
        val cell = (this.size.minDimension / cells) * scale
        val originX = (this.size.width - cell * cells) / 2f
        // A gentle bob for the happy ones, half a pixel of life — read here,
        // in the draw phase, so it repaints without recomposing.
        val bob = if (bobActive) abs(phase.value - 1f) else 0f
        val originY = (this.size.height - cell * cells) / 2f + (bob * cell * 0.5f)

        fun draw(list: List<PetCell>, yOffsetRows: Int) {
            for (c in list) {
                drawRect(
                    color = c.color,
                    topLeft = Offset(originX + c.x * cell, originY + (c.y + yOffsetRows) * cell),
                    size = Size(cell + 0.5f, cell + 0.5f),
                )
            }
        }

        draw(bodyCells, 0)
        if (crownCells != null) draw(crownCells, -1)
    }
}

private class PetCell(val x: Int, val y: Int, val color: Color)

private fun resolveCells(rows: List<String>, body: Color, shade: Color): List<PetCell> {
    val out = ArrayList<PetCell>()
    rows.forEachIndexed { y, row ->
        row.forEachIndexed { x, ch ->
            val color = when (ch) {
                'o' -> Color(0xFF1A1721)
                'b' -> body
                'B' -> shade
                'w' -> Color(0xFFF2F0F8)
                'p' -> Color(0xFFFF8FA3)
                'e' -> Color(0xFF17151F)
                'y' -> Color(0xFFFCCE09)
                else -> null
            } ?: return@forEachIndexed
            out += PetCell(x, y, color)
        }
    }
    return out
}

File: ios/Yappy/Components/PixelPet.swift
import SwiftUI
import UIKit

/// The group pet: a pixel creature, drawn from character grids, animated by
/// flipping between two frames the way a Tamagotchi did.
///
/// Sprites are code, not assets — each frame is sixteen strings of sixteen
/// characters, one character per pixel, so the art is diffable, palette-swaps
/// with the group's identity colour, and never ships a PNG. The vocabulary:
///
///   .  transparent      o  outline          b  body (identity colour)
///   B  body shade       w  white            p  pink (tongue, inner ear)
///   e  eye              y  brand yellow (sparkles, crown)
///
/// Species is derived from the conversation id: half the world's groups get a
/// dog, half a cat, nobody chooses and everybody agrees.
///
/// Kept in step with android/.../ui/components/PixelPet.kt — the grids are
/// copied verbatim, and the species derivation reproduces Kotlin's hash so both
/// platforms draw the same creature for the same group.

enum PetSpecies {
    case dog
    case cat
}

/// Kotlin/Java `String.hashCode()`: h = 31·h + c over UTF-16 code units, with
/// Int32 overflow wrapping. Android picks the species from its own
/// `String.hashCode()`, so parity means reproducing that exact arithmetic —
/// Swift's `hashValue` is per-process seeded and useless here.
func javaHashCode(_ s: String) -> Int32 {
    var h: Int32 = 0
    for unit in s.utf16 {
        h = 31 &* h &+ Int32(unit)
    }
    return h
}

func petSpecies(conversationId: String) -> PetSpecies {
    (javaHashCode(conversationId) & 1) == 0 ? .dog : .cat
}

// ─── Sprites ─────────────────────────────────────────────────────────────────

private let eggFrames: [[String]] = [
    [
        "................",
        "................",
        "......oooo......",
        ".....obbbbo.....",
        "....obbwbbbo....",
        "...obbwbbbbbo...",
        "...obbbbbbbbo...",
        "..obbbbbbbbbbo..",
        "..obbBbbbbBbbo..",
        "..obbbbbbbbbbo..",
        "..obBbbbbbbBbo..",
        "...obbbbbbbbo...",
        "....oobbbboo....",
        "......oooo......",
        "................",
        "................",
    ],
    [
        "................",
        "................",
        "................",
        "......oooo......",
        ".....obbbbo.....",
        "....obbwbbbo....",
        "...obbwbbbbbo...",
        "...obbbbbbbbo...",
        "..obbbbbbbbbbo..",
        "..obbBbbbbBbbo..",
        "..obbbbbbbbbbo..",
        "..obBbbbbbbBbo..",
        "...obbbbbbbbo...",
        ".....oooooo.....",
        "................",
        "................",
    ],
]

// Dog: floppy ears, big muzzle. Frame two lifts the ears and wags.
private func dogFrames(mood: String) -> [[String]] {
    // Rows 10-11 are the mouth region, swapped per mood.
    func body(_ earUp: Bool, _ mouthA: String, _ mouthB: String) -> [String] {
        let e1 = earUp ? "..oo........oo.." : "................"
        let e2 = earUp ? ".obbo......obbo." : "..oo........oo.."
        let e3 = earUp ? ".obBbo....obBbo." : ".obbo......obbo."
        return [
            "................",
            e1,
            e2,
            e3,
            ".obBbooooooBbbo.",
            ".obbobbbbbbobbo.",
            "..oobbbbbbbboo..",
            "..obbebbbbebbo..",
            "..obbbbbbbbbbo..",
            "..obbbBooBbbbo..",
            mouthA,
            mouthB,
            "...obbbbbbbbo...",
            "....oooooooo....",
            "................",
            "................",
        ]
    }
    switch mood {
    case "happy":
        return [
            body(false, "..obbboppobbbo..", "...obbboppbbo..."),
            body(true, "..obbboppobbbo..", "...obbbbppbo...."),
        ]
    case "hungry":
        return [
            body(false, "..obbbboobbbbo..", "...obbbbbbbbo..."),
            body(false, "..obbboooobbbo..", "...obbbbbbbbo..."),
        ]
    default: // sad
        return [
            body(false, "..obbbboobbbbo..", "...obbboobbbo..."),
            body(false, "..obbbboobbbbo..", "...obbboobbbo..."),
        ]
    }
}

// Cat: pointed ears, small mouth. Frame two flicks an ear and the tail.
private func catFrames(mood: String) -> [[String]] {
    func body(_ flick: Bool, _ mouthA: String, _ mouthB: String) -> [String] {
        let e1 = flick ? "..o..........o.." : "..o.........o..."
        let e2 = flick ? "..opo......opo.." : "..opo......opo.."
        return [
            "................",
            e1,
            e2,
            "..obpo....obpo..",
            "..obboooooobbo..",
            ".obbbbbbbbbbbbo.",
            ".obebbbbbbbebbo.",
            ".obbbbbbbbbbbbo.",
            "..obbBwbbwBbbo..",
            mouthA,
            mouthB,
            "..obbbbbbbbbbo..",
            "...obbbbbbbbo...",
            "....oooooooo....",
            "................",
            "................",
        ]
    }
    switch mood {
    case "happy":
        return [
            body(false, "..obbbopbobbbo..", "...obbbbbbbbo..."),
            body(true, "..obbbobpobbbo..", "...obbbbbbbbo..."),
        ]
    case "hungry":
        return [
            body(false, "..obbbboobbbbo..", "...obbbbbbbbo..."),
            body(true, "..obbbboobbbbo..", "...obbbbbbbbo..."),
        ]
    default: // sad
        return [
            body(false, "..obbboBBobbbo..", "...obbbbbbbbo..."),
            body(false, "..obbboBBobbbo..", "...obbbbbbbbo..."),
        ]
    }
}

// Wandered off: an empty spot — footprints trailing away and a question.
private let goneFrames: [[String]] = [
    [
        "................",
        "......ww........",
        ".....w..w.......",
        "........w.......",
        ".......w........",
        ".......w........",
        "................",
        ".......w........",
        "................",
        "..BB............",
        "..BB....BB......",
        "........BB......",
        "............BB..",
        "............BB..",
        "................",
        "................",
    ],
]

// Elder crown, drawn over the sprite's head rows.
private let crownRows: [String] = [
    "....y..y..y.....",
    "....yyyyyy......",
]

// ─── Rendering ───────────────────────────────────────────────────────────────

/// - Parameters:
///   - conversationId: decides species and the body colour.
///   - stage: egg | baby | kid | grown | elder — size and form.
///   - mood: happy | hungry | sad | gone — expression and tempo.
struct PixelPet: View {
    let conversationId: String
    let stage: String
    let mood: String
    let size: CGFloat
    var animated: Bool = true

    var body: some View {
        TimelineView(.animation(minimumInterval: 1.0 / 30.0, paused: !animated)) { timeline in
            sprite(at: timeline.date)
        }
        .frame(width: size, height: size)
    }

    private func sprite(at date: Date) -> some View {
        let bodyColor = colorForId(conversationId)
        let shade = shadeOf(bodyColor)

        let frames: [[String]]
        if mood == "gone" {
            frames = goneFrames
        } else if stage == "egg" {
            frames = eggFrames
        } else if petSpecies(conversationId: conversationId) == .dog {
            frames = dogFrames(mood: mood)
        } else {
            frames = catFrames(mood: mood)
        }

        // Sad pets breathe slowly; happy ones can barely sit still.
        let periodSeconds: Double
        switch mood {
        case "happy": periodSeconds = 0.380
        case "hungry": periodSeconds = 0.650
        default: periodSeconds = 0.900
        }

        // The phase runs 0 → 2 and wraps, the same shape as Android's repeated
        // tween: the frame flips at 1, the bob is the distance from 1.
        let phase = animated
            ? date.timeIntervalSinceReferenceDate
                .truncatingRemainder(dividingBy: periodSeconds * 2) / periodSeconds
            : 0
        let frame = (animated && frames.count > 1 && phase >= 1) ? 1 : 0
        // A gentle bob for the happy ones, half a pixel of life.
        let bob = (animated && mood == "happy" && stage != "egg") ? abs(phase - 1) : 0

        let grid = frames[frame]
        // Babies are the same creature, smaller in the same box.
        let scale: CGFloat
        switch stage {
        case "baby": scale = 0.72
        case "kid": scale = 0.88
        default: scale = 1
        }

        return Canvas { context, canvasSize in
            let cells = 16
            let cell = (min(canvasSize.width, canvasSize.height) / CGFloat(cells)) * scale
            let originX = (canvasSize.width - cell * CGFloat(cells)) / 2
            let originY = (canvasSize.height - cell * CGFloat(cells)) / 2 + CGFloat(bob) * cell * 0.5

            func colorFor(_ ch: Character) -> Color? {
                switch ch {
                case "o": return Color(hex: 0x1A1721)
                case "b": return bodyColor
                case "B": return shade
                case "w": return Color(hex: 0xF2F0F8)
                case "p": return Color(hex: 0xFF8FA3)
                case "e": return Color(hex: 0x17151F)
                case "y": return Color(hex: 0xFCCE09)
                default: return nil
                }
            }

            func drawGrid(_ rows: [String], yOffsetRows: Int = 0) {
                for (y, row) in rows.enumerated() {
                    for (x, ch) in row.enumerated() {
                        guard let color = colorFor(ch) else { continue }
                        // The half-point overlap closes the hairline seams a
                        // non-integral cell size would otherwise leave.
                        let rect = CGRect(
                            x: originX + CGFloat(x) * cell,
                            y: originY + CGFloat(y + yOffsetRows) * cell,
                            width: cell + 0.5,
                            height: cell + 0.5
                        )
                        context.fill(Path(rect), with: .color(color))
                    }
                }
            }

            drawGrid(grid)
            if stage == "elder" && mood != "gone" {
                drawGrid(crownRows, yOffsetRows: -1)
            }
        }
    }

    /// The body colour at 72% — the same factor Android multiplies in.
    private func shadeOf(_ color: Color) -> Color {
        var r: CGFloat = 0, g: CGFloat = 0, b: CGFloat = 0, a: CGFloat = 0
        _ = UIColor(color).getRed(&r, green: &g, blue: &b, alpha: &a)
        return Color(.sRGB, red: Double(r) * 0.72, green: Double(g) * 0.72, blue: Double(b) * 0.72, opacity: 1)
    }
}
